File: bindings/go/oprf_test.go
package quipu

import (
	"bytes"
	"errors"
	"testing"
)

// Smoke tests offline del FFI VOPRF (no requieren servidor). El round-trip
// válido completo se cubre en scripts/oprf-e2e.sh.

func TestVoprfBlindSizes(t *testing.T) {
	state, blinded, err := VoprfBlind([]byte("password"))
	if err != nil {
		t.Fatal(err)
	}
	if len(state) != 64 {
		t.Fatalf("state = %d B, want 64", len(state))
	}
	if len(blinded) != 32 {
		t.Fatalf("blinded = %d B, want 32", len(blinded))
	}
}

func TestVoprfBlindIsRandomized(t *testing.T) {
	_, b1, _ := VoprfBlind([]byte("password"))
	_, b2, _ := VoprfBlind([]byte("password"))
	if bytes.Equal(b1, b2) {
		t.Fatal("dos cegados de la misma password deben diferir (aleatorios)")
	}
}

func TestVoprfFinalizeRejectsBogusProof(t *testing.T) {
	state, _, err := VoprfBlind([]byte("password"))
	if err != nil {
		t.Fatal(err)
	}
	// Evaluación/prueba/clave inventadas: la prueba DLEQ no valida -> ErrAuth.
	_, err = VoprfFinalize([]byte("password"), state, make([]byte, 32), make([]byte, 64), make([]byte, 32))
	if !errors.Is(err, ErrAuth) {
		t.Fatalf("esperaba ErrAuth, fue %v", err)
	}
}

File: bindings/node/test/oprf.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import * as quipu from '../src/index.js';

// Smoke tests offline del FFI VOPRF (no requieren servidor). El round-trip
// válido completo se cubre en scripts/oprf-e2e.sh.

test('voprfBlind returns state(64) and blinded(32)', () => {
  const { state, blinded } = quipu.voprfBlind(Buffer.from('password'));
  assert.equal(state.length, 64);
  assert.equal(blinded.length, 32);
});

test('voprfBlind is randomized', () => {
  const a = quipu.voprfBlind(Buffer.from('password')).blinded;
  const b = quipu.voprfBlind(Buffer.from('password')).blinded;
  assert.ok(!a.equals(b));
});

test('voprfFinalize rejects a bogus proof -> AUTH', () => {
  const { state } = quipu.voprfBlind(Buffer.from('password'));
  assert.throws(
    () => quipu.voprfFinalize(Buffer.from('password'), state, Buffer.alloc(32), Buffer.alloc(64), Buffer.alloc(32)),
    (e) => {
      assert.ok(e instanceof quipu.QuipuError);
      assert.equal(e.code, 'AUTH');
      return true;
    },
  );
});

// La salida de RFC 9497 es el SHA-512 entero: 64 B, no 32. Antes de la
// conformidad eran 32; este test caza cualquier vuelta atras.
test('oprfHarden devuelve 64 B contra un servidor real', { skip: !process.env.QUIPU_OPRF_API_KEY }, async () => {
  const base = process.env.QUIPU_OPRF_URL ?? 'http://127.0.0.1:8791';
  const pub = Buffer.from((await (await fetch(base + '/v1/public-key')).json()).public_key, 'hex');
  const secret = await quipu.oprfHarden({
    baseUrl: base,
    apiKey: process.env.QUIPU_OPRF_API_KEY,
    password: Buffer.from('rfc9497', 'utf8'),
    serverPublicKey: pub,
  });
  assert.equal(secret.length, 64);
});
